lib/types-simple.nix

```diff
@@ -0,0 +1,457 @@
+        else if t.variant == variants.recursive then
+          mapAndFilter (_: el: t.nested el) (t.each v)
+        # there’s exactly one tagged value, so check that
+        else if t.variant == variants.sum       then
+              # we already tested length == 1 in .check
+          let alt = builtins.head (builtins.attrNames v);
+          in t.alts.${alt} v.${alt}
+        # check each field according to its type
+        # optional missing fields of course always pass the check
+        else if t.variant == variants.product   then
+          mapAndFilter (n: f: if v ? ${n} then f v.${n} else ok)
+                       (t.req // t.opt)
+        # if the value fails the check for each type it can have,
+        # we throw an error; if one check succeeds the union is satisfied
+        else if t.variant == variants.union     then
+          # unions are awkward, the type checker can’t do much here
+          if lib.all (res: res != ok) (map (f: f v) t.altList)
+          then { should = t.description; val = v; }
+          else ok
+        else unreachable;
+    # cata only has “two arguments”, giving it a Value as third
+    # argument “morphs” the `a` in alg to (Value -> Errors);
+    # of course we could curry t and v away,
+    # but just `cata alg` would be very confusing ;)
+    in t: v: cata alg t v;
+
+
+  ## -- TYPE SETUP STUFF --
+
+  mkBaseType = {
+    # unique name (for matching on the type)
+    name,
+    # the (displayable) type description
+    description,
+    # a function to check the outermost type, given a value (Val -> Bool)
+    # TODO: this is value-specific, maybe this should be inside the type checker
+    # logic instead of the type definiton? There’s some repetition.
+    check,
+    # the variant of this type
+    variant,
+    # extra fields belonging to the variant
+    extraFields
+  }: { inherit name description check variant; } // extraFields;
+
+  mkScalar = { name, description, check }: mkBaseType {
+    inherit name description check;
+    variant = variants.scalar;
+    extraFields = {};
+  };
+
+  mkRecursive = { name, description, check,
+    # return all children for a value of this type T t,
+    # give each child (of type t) a displayable name.
+    # (T t -> Map Name t)
+    each,
+    # The nested value t of the type functor
+    nested
+  }: mkBaseType {
+    inherit name description check;
+    variant = variants.recursive;
+    extraFields = { inherit each nested; };
+  };
+
+
+  ## -- TYPES --
+
+  # the type with no inhabitants (kind of useless …)
+  void = mkScalar {
+    name = "void";
+    description = "void";
+    # there are no values of type void
+    check = lib.const false;
+  };
+
+  # the any type, every value is an inhabitant
+  # it basically turns off the type system, use with care
+  any = mkScalar {
+    name = "any";
+    description = "any type";
+    check = lib.const true;
+  };
+
+  # the type with exactly one inhabitant
+  unit = mkScalar {
```

Comment:
Why do we need this type?

Reply:
Similar to `void` this doesn’t seem useful at first, but it’s even more elementar than `bool` in a sense; e.g. the `void` you find in languages like `C(++)` is really a `unit`.

A great use-case is for sum types:

```
let mode = sum { off = unit; level = int; };
in checkType mode { off = {}; };
```

which corresponds to the Haskell type:

```
data Mode = Off | Level Int
```